libpromises/dbm_lmdb.c

```diff
@@ -223,11 +227,289 @@ static int LmdbEnvOpen(
+        {
+            /* The corruption has already been handled. This process should
+             * just die because we have no way to return to the point where
+             * it would just open the new (repaired) LMDB file. */
+            handle_corruption = false;
+        }
+        SharedUnlockFile(fd_tstamp); /* closes fd_tstamp (TODO: fix) */
+        fd_tstamp = safe_open(tstamp_file, O_CREAT|O_RDWR);
+    }
+    else
+    {
+        /* should never happen (we tried to wait), but if it does, just log an
+         * error and keep going */
+        Log(LOG_LEVEL_ERR,
+            "Failed to get shared lock for the repair timestamp of the '%s' DB",
+            lmdb_file);
+    }
+
+    if (!handle_corruption)
+    {
+        /* Just clean after ourselves and terminate the process. */
+        ThreadUnlock(cft_db_corruption_lock);
+        close(fd_db_lock);
+        close(fd_tstamp);
+        free(db_lock_file);
+        free(tstamp_file);
+
+        /* TODO: no need to _exit(), we should be able to just make sure
+         * atexit stuff doesn't touch the repaired DB and exit(1). */
+
+        /* Avoid running atexit handlers that would mess with the LMDB files
+         * including the corrupted one and, worse, potentially even the new repaired
+         * one. */
+        _exit(1);
+    }
+
+    /* HERE is a window for some other process to do the repair between when we
+     * checked the timestamp using the shared lock above and the attempt to get
+     * the exclusive lock right below. However, this is detected by checking the
+     * contents of the timestamp file again below, while holding the EXCLUSIVE
+     * lock. */
+
+    ret = ExclusiveLockFile(fd_tstamp, true);
+    if (ret == -1)
+    {
+        /* should never happen (we tried to wait), but if it does, just
+         * terminate because doing the repair without the lock could be
+         * disasterous */
+        Log(LOG_LEVEL_ERR,
+            "Failed to get shared lock for the repair timestamp of the '%s' DB",
+            lmdb_file);
+
+        ThreadUnlock(cft_db_corruption_lock);
+        close(fd_db_lock);
+        close(fd_tstamp);
+        free(db_lock_file);
+        free(tstamp_file);
+
+        /* TODO: no need to _exit(), we should be able to just make sure
+         * atexit stuff doesn't touch the repaired DB and exit(1). */
+
+        /* Avoid running atexit handlers that would mess with the LMDB files
+         * including the corrupted one and, worse, potentially even the new
+         * repaired one. */
+        _exit(1);
+    }
+
+    /* Cleared to resolve the corruption. */
+
+    /* 1. Acquire the lock for the DB to prevent more processes trying to use
+     *    it while it is corrupted (wait till the lock is available). */
+    while (ExclusiveLockFile(fd_db_lock, false) == -1)
+    {
+        /* busy wait to do the logging */
+        Log(LOG_LEVEL_INFO, "Waiting for the lock on the '%s' DB",
+            lmdb_file);
+        sleep(1);
```

Comment:
No timeout of giving up?
Log every second? Seems a bit much maybe.

Reply:
This lock is only held while a DB is being opened. That can lead to this function being called, but it gives up the lock ASAP. So, in general, this lock should not be held by any process for any significant amount of time. And thus it is desired to logging in this way, I think. And for the same reason, this should not give up. It should never have to do so.

This may be refined in the future, but I'd like to make things a bit aggressive here first so that we discover faults in our assumptions ASAP.